### PR TITLE
Align config/output configuration

### PR DESCRIPTION
**Pull request by rrahn**
fixes seqan/product_backlog#180

Replaces the align_cfg::result with the align_cfg::output. 
The align_cfg::result is still needed in order to set the score type which will be fixed by another issue and is not part of this PR. So removing the result option will be done later when the score type is also changed to the new configuration.

### CODE REVIEW

File: include/seqan3/alignment/configuration/align_config_output.hpp
Comment (Irallia) [inline include/seqan3/alignment/configuration/align_config_output.hpp:40]: My English is not the best but I would leave out the 'then'.
```suggestion
 * This option forces the alignment to compute and output the score. If this option is not set in the alignment
 * configuration, accessing the score via the seqan3::alignment_result object is forbidden and will lead to a compile 
 * time error.
```

File: include/seqan3/alignment/configuration/align_config_output.hpp
Comment (Irallia) [inline include/seqan3/alignment/configuration/align_config_output.hpp:73]:
```suggestion
 * The end positions denote the end of the alignment within the original sequences, i.e. the position behind
 * the last aligned character.
```
or
```suggestion
 * The end positions denote the end of the alignment within the original sequences, i.e. the positions behind
 * the last aligned characters.
```

File: include/seqan3/alignment/configuration/align_config_output.hpp
Comment (Irallia) [inline include/seqan3/alignment/configuration/align_config_output.hpp:109]: same here
```suggestion
 * The begin positions denote the begin of the alignment within the original sequences, i.e. the positions of the
 * first aligned characters.
```

File: include/seqan3/alignment/configuration/align_config_output.hpp
Comment (Irallia) [inline include/seqan3/alignment/configuration/align_config_output.hpp:111]:
```suggestion
 * If this option is not set in the alignment configuration, accessing the begin positions via the
```

File: include/seqan3/alignment/configuration/align_config_output.hpp
Comment (Irallia) [inline include/seqan3/alignment/configuration/align_config_output.hpp:143]:
```suggestion
 * If this option is not set in the alignment configuration, accessing the alignment via the
```

File: include/seqan3/alignment/configuration/align_config_output.hpp
Comment (Irallia) [inline include/seqan3/alignment/configuration/align_config_output.hpp:175]:
```suggestion
 * If this option is not set in the alignment configuration, accessing the id of the first sequence via the
```

File: include/seqan3/alignment/configuration/align_config_output.hpp
Comment (Irallia) [inline include/seqan3/alignment/configuration/align_config_output.hpp:207]:
```suggestion
 * If this option is not set in the alignment configuration, accessing the id of the second sequence via the
```

File: include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp
Comment (Irallia) [inline include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp:1063]:
```suggestion
        // configuration traits and not in the edit distance traits.
```

File: test/unit/alignment/configuration/align_config_common_test.cpp
Comment (Irallia) [inline test/unit/alignment/configuration/align_config_common_test.cpp:56]: When does this happen? Is this part of this PR?

File: include/seqan3/alignment/configuration/align_config_output.hpp
Comment (rrahn) [reply include/seqan3/alignment/configuration/align_config_output.hpp:40]: Both versions are grammatically correct. The 'then' however stresses the second part more which I find important here. But no strong feelings. If you still want me to change it, I'll do it ;).

File: test/unit/alignment/configuration/align_config_common_test.cpp
Comment (rrahn) [reply test/unit/alignment/configuration/align_config_common_test.cpp:56]: This happens in the first 5 commits. Each output option is a separate commit and updates this value.

File: include/seqan3/alignment/configuration/align_config_output.hpp
Comment (Irallia) [inline include/seqan3/alignment/configuration/align_config_output.hpp:75]: We missed one:
```suggestion
 * If this option is not set in the alignment configuration, then accessing the end positions via the
```

**Comment by rrahn** [comment]
Great, thanks. @smehringer I will first rebase and then ping you again once the commits are clean again.

**Comment by rrahn** [comment]
@smehringer this is now ready for second review. Thanks a lot.

**Comment by rrahn** [comment]
@smehringer ping pong

File: include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp
Comment (smehringer) [inline include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp:1049]:
```suggestion
        // Note, that they might be activated even if the user did not configure them, but in order to
```

File: include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp
Comment (smehringer) [inline include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp:1059]: What if `if constexpr (compute_end_positions)` before evaluated to false and the end positions have not been properly initialised?

File: include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp
Comment (smehringer) [inline include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp:1079]: I don't understand the comment in this context. Where do you compute the begin and end position "if not cashed already"? It seems you are just using the `cashed_[..]` variables without checking?

File: include/seqan3/alignment/pairwise/detail/type_traits.hpp
Comment (smehringer) [inline include/seqan3/alignment/pairwise/detail/type_traits.hpp:141]:
```suggestion
    //!\brief The selected result type.
```

File: test/unit/alignment/pairwise/align_pairwise_test.cpp
Comment (smehringer) [inline test/unit/alignment/pairwise/align_pairwise_test.cpp:181]: This also seems to be a TODO. Can you maybe add all these lines to the respective issue so we don't forget about them?

File: test/unit/alignment/pairwise/align_pairwise_test.cpp
Comment (smehringer) [inline test/unit/alignment/pairwise/align_pairwise_test.cpp:263]: same here

File: test/unit/alignment/pairwise/align_result_selector_test.cpp
Comment (smehringer) [inline test/unit/alignment/pairwise/align_result_selector_test.cpp:81]: here

File: include/seqan3/alignment/all.hpp
Comment (smehringer) [inline include/seqan3/alignment/all.hpp:129]:
```suggestion
 * aligned. For example, the begin and end positions of a global alignment encompass the entire sequences, 
 * but in case of a local alignment the
```
The sentence before was a little confusing and I think didn't bring in extra information.

File: include/seqan3/alignment/all.hpp
Comment (smehringer) [inline include/seqan3/alignment/all.hpp:134]:
```suggestion
 * To obtain the actual alignment the option seqan3::align_cfg::output_alignment has to be specified.
 * The options can be combiend with each other in order to customise the alignment algorithm and the respective output of the
```
How does the output customise the alignment? It can only customise the alignment algorithm, no?
I think you use "alignment" as in "the deed of aligning", but also for the resulting trace. That's a little confusing.

File: include/seqan3/alignment/all.hpp
Comment (smehringer) [inline include/seqan3/alignment/all.hpp:140]: Hm isn't that intuitive? We compute everything but then wouldn't let people access those information? 
Or did I understand this sentence wrong?

File: include/seqan3/alignment/pairwise/detail/type_traits.hpp
Comment (rrahn) [reply include/seqan3/alignment/pairwise/detail/type_traits.hpp:186]: It is a separate issue that I will work on afterwards.

File: include/seqan3/alignment/configuration/align_config_output.hpp
Comment (rrahn) [reply include/seqan3/alignment/configuration/align_config_output.hpp:26]: No. In case we have a simple configuration we decided to call the class `_tag` and the instance of it without the tag suffix. This has nothing to do with detail or non-detail code.

File: include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp
Comment (rrahn) [reply include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp:1059]: Then, someone configured the edit distance wrong. This is an internal dependency. I will add a static_assert to make sure that this is safe at compile time.

File: include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp
Comment (rrahn) [reply include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp:1079]: it is an old comment. I'll remove it

File: test/unit/alignment/pairwise/align_pairwise_test.cpp
Comment (rrahn) [reply test/unit/alignment/pairwise/align_pairwise_test.cpp:181]: This won't be necessary, because when the PR with the score_type is through the result config will be deleted and all the tests wouldn't compile.

File: test/unit/alignment/pairwise/align_pairwise_test.cpp
Comment (rrahn) [reply test/unit/alignment/pairwise/align_pairwise_test.cpp:263]: like above

File: include/seqan3/alignment/all.hpp
Comment (rrahn) [reply include/seqan3/alignment/all.hpp:134]: Maybe a sentence is missing

File: include/seqan3/alignment/all.hpp
Comment (rrahn) [reply include/seqan3/alignment/all.hpp:140]: maybe a sentence is missing

File: include/seqan3/alignment/configuration/align_config_output.hpp
Comment (smehringer) [reply include/seqan3/alignment/configuration/align_config_output.hpp:26]: > In case we have a simple configuration we decided to call the class _tag and the instance of it without the tag suffix

Yes absolutely

>  This has nothing to do with detail or non-detail code.

I still think so, at least it is that way in all the other condifurations:
`seqan3::detail::method_local_tag` for `seqan3::align_cfg::method_local`
`seqan3::detail::vectorised_tag` for `seqan3::align_cfg::vectorised`
`seqan3::detail::output_query_id_tag` for `seqan3::search_cfg::output_query_id`
...

File: include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp
Comment (smehringer) [inline include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp:1059]:
```suggestion
            static_assert(compute_end_positions, "End positions required to compute the begin positions.");
```
?

File: test/unit/alignment/pairwise/align_pairwise_test.cpp
Comment (smehringer) [reply test/unit/alignment/pairwise/align_pairwise_test.cpp:181]: :smile: well thats true. You have many TODO comments in this PR though, so I thought you deliberately did this..

**Review by smehringer — CHANGES_REQUESTED** [review]

File: test/unit/alignment/pairwise/align_pairwise_test.cpp
Comment (rrahn) [reply test/unit/alignment/pairwise/align_pairwise_test.cpp:181]: It was more or less the places that I had to touch but couldn't resolve completely so I marked them. But not all places seem to be affected by my change so I didn't see this.

File: include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp
Comment (rrahn) [reply include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp:1059]: huh interesting 🤓

File: include/seqan3/alignment/configuration/align_config_output.hpp
Comment (rrahn) [reply include/seqan3/alignment/configuration/align_config_output.hpp:26]: No 😱 they should not be in detail. Otherwise, how is the user supposed to call get on it. `seqan3::get<seqan3::align_cfg::method_local>` is not working and if the tag is in detail namespace, then he can't see the user documentation.

Comment (smehringer) [reply include/seqan3/alignment/configuration/align_config_output.hpp:26]: Haha well somewhere this went wrong then. I always thought that all the special configuration functionality except the `|` was rather for us internally anyway. Will you open an issue for this? I guess it should be in the current release then.

**Review by smehringer — APPROVED** [review]

File: include/seqan3/alignment/configuration/align_config_output.hpp
Comment (rrahn) [reply include/seqan3/alignment/configuration/align_config_output.hpp:26]: > Haha well somewhere this went wrong then.

Yes, by neither reading the user story nor the config design document. 😒

**Comment by smehringer** [comment]
See comment in Gitter before you merge

**Pull request merged**
